### DISCUSSION TITLE
Tell us how you are using Flows!

### DISCUSSION THREAD

**Discussion by saswatds** [postmanlabs/postman-flows#373] [General]
Postman Flows has been in beta for a while and now seems a good time to ask everyone how they are using flows in their regular workflow. Please drop in a short comment about how flows has solved your problem or if you see the potential.

**Comment by yatw** [comment]
I am using flow to set up the initial admin users in my application. I will call the create user API and call assign user role API.
2 issues I am seeing with flow is 
1. For the create user API, we are unable to specify different API input body for each block to create many users.
2. To do the same in different environment, we have to manually change the environment in each block
Otherwise flow is a great concept.

**Reply by saswatds** [reply]
@yatw This seems like a use case that definitely something flows should solve instantly. Let's think about a few solutions here and we will implement them immediately to solve your use-case.

> we are unable to specify different API input body for each block to create many users 
I have a few question - 
* Do you wish to have some kind of looping behaviour here for bulk creation? 
* In what way all the data for the create user API stored? Is it a file, a large JSON array? 

> we have to manually change the environment in each block
* We had an implementation where changing the selected environment for one request changed it for all similar request. I believe this is not being helpful to you
* Does being able to select an environment before execution solve your use-case?

**Reply by yatw** [reply]
yes I want a looping behavior for each of the admin user. 
The API input data is a json object like 
`{
    "email": "yatw1@gmail.com",
    "account_area_units": "m2",
    "account_currency": "USD",
    "default_org_id": "1"
}`
for each user

As for changing the environment, we are looking for something simple like being able to select an environment once before flow execution

This is what I am trying, still working on how to pass in the api input to the request. Please kindly give us suggestion. Thanks
<img width="719" alt="Screen Shot 2022-08-16 at 17 25 11" src="https://user-images.githubusercontent.com/19565177/185007618-d5b75eed-44a4-4b0e-aefe-d11f5c4c6b99.png">

**Reply by yatw** [reply]
<img width="747" alt="Screen Shot 2022-08-17 at 11 16 03" src="https://user-images.githubusercontent.com/19565177/185213031-ebbcc53e-6e46-4c1b-9e17-adb204549b6d.png">
it took me sometime but I figured out how to create data into the api request

**Reply by miquecg** [reply]
I'm interested in using Postman Flows but I have not yet a clear idea if it would be possible to override a request body. @yatw It looks like your Create Data step just fills in all relevant fields needed for `Create a user` request, right?

@saswatds In the same vein, are data generators like `$randomPhoneNumber` available as in pre-request scripts?

My use case is a PATCH request with possible different bodies and response outcomes (incremental updates of user info), so I need to write different tests for each one. If I can write tests in blocks that will be fine, but if I need to have them in requests from the collection, I'd basically end up with what I already have, duplicating the exact same request with slightly different inputs and tests. Having blocks with different tests chained together poses its own problems also, mainly visual clutter in the workspace, specially without any support from lables or custom names for blocks. Duplication can be reduced looping over more complex constructs (I'm not aware of it), let's say one block is like: ( Data Input > Send Request > Test ) and is executed in five iterations, each of one has different data and different assertions. That would be ideal :)

**Reply by yatw** [reply]
I also have the same question too. So what I did now is for Create a user api, I make the request body a variable like this.
<img width="411" alt="Screen Shot 2022-08-18 at 09 47 25" src="https://user-images.githubusercontent.com/19565177/185450330-162f2e92-e4e6-482f-b7b7-f693a95a9f63.png">
Then in the flow, send request block, I pass in the body as variable
<img width="625" alt="Screen Shot 2022-08-18 at 09 48 31" src="https://user-images.githubusercontent.com/19565177/185450551-42b5fc05-f4dd-4384-b5aa-8622074ce56f.png">
I am also wondering if there is a better way, because API request body needs to be change to variable to use flow

**Comment by Ariemeth** [comment]
Currently I've used Flows to build flows to create and delete test users from our system.  Next up is generating various types of group relationships for those test users.

My way of handling the need for multiple user accounts is having a list of 100 strings that I loop over to create N users
![image](https://user-images.githubusercontent.com/541023/185176069-0a9de7ed-5825-4363-b705-b1c39cf0d900.png)

Really looking forward to being able to convert a number to a string instead of having to create data lists, especially when that list is just 
"1", "2", etc...

**Reply by saswatds** [reply]
Hey @Ariemeth there is also a block called `Loop N times` which take a number. Will that simplify this use-case?

**Reply by Ariemeth** [reply]
No, because there is no convert number to string ability.

**Reply by saswatds** [reply]
I can add a simple casting block thought the existing `String` block. That should do it right?

**Reply by Ariemeth** [reply]
That would be awesome!

**Reply by Ariemeth** [reply]
Just previewed test user and group creation to my QA team and they are super excited to use them.

**Comment by NathanAP** [comment]
I'm using flows to create a guide for our API. I'm in the learning process of how to organize things but I think it could have a lot of good uses for it. Seems that it has a potential to demonstrate how an application flow works in a simple way! Pretty interesting tool that I might study more and more.

Looking forward for these:
- be able to name blocks and change its colors
- be able to lock examples to each block and/or flows
- be able to create "areas" that I could identify as "this area is responsable for making login, while that one is responsable for create a new user"

**Reply by saswatds** [reply]
Hey @NathanAP thanks for trying out flows. 
> be able to name blocks and change its colors

This is on our backlog

> be able to lock examples to each block and/or flows 

I would like to understand this a bit better.

> be able to create "areas" that I could identify as "this area is responsable for making login, while that one is responsable for create a new user"

This is a good idea. I will add it to our internal idea tracker.

**Reply by NathanAP** [reply]
> Hey @NathanAP thanks for trying out flows.
> 
> > be able to name blocks and change its colors
> 
> This is on our backlog
> 
> > be able to lock examples to each block and/or flows
> 
> I would like to understand this a bit better.
> 
> > be able to create "areas" that I could identify as "this area is responsable for making login, while that one is responsable for create a new user"
> 
> This is a good idea. I will add it to our internal idea tracker.

About the second point, maybe I could use Flows differently, but I'll give an example: lets say that I have a request that can only be made by an admin user. If I change my login example to another user, that request in my flow would break it. My idea is that I could just mark it as locked or assigned to a Flow, so my team won't change it as well. Its not really important or urgent, but it could avoid some situations.

In the current version, I'm naming my examples as "FLOW", so I know that I cannot change that example to not break things in my Flows.

Also, I'm really missing the chance to create variables in enviroment through Flows' request. My requests depends on them, and right know I feel that I need to create two kind of collections, one for tests and another for flows. How do you suggest us to manage and organize flows requests vs default/testing requests?

**Comment by lHuKKal** [comment]
Hello!

I am using the flow to test api in our 2 systems and i really like the flow =) Just one question, will you implementing import/export of the flow in future?

**Comment by ShawnFUB** [comment]
A use case I'm interested in but don't see how to implement is using Flow for automated testing of entire API workflows rather than individual requests. E.g.:

- GET /people
- Assert user "John Doe" **is not in** the list of people.
- POST /people - Create "John Doe"
- GET /people
- Assert "John Doe" **is in** the list of people.
- PATCH /people/id change "John Doe" details
- GET /people/id
- Assert "John Doe" **was updated** correctly.
- DELETE /people/id
- GET /people
- Assert "John Doe" **is not in** the list of people.

Along with the ability to clean up the test user if any intermediate step fails after creating the user. Flows is _almost_ there, but I don't see any way of making assertions as part of the flow - Postman assumes that they are part of the Request itself, but I can't easily write tests for GET /people to both assert that the user is there and (later) isn't there. Duplicating the Request with different Test scripts works, but feels hacky. Adding an if/else based on environment variables also works, but again, feels hacky. The test assertion only really makes sense across multiple Requests and not tied to any one Request.

Being able to elevate test assertions to the Flow itself and get them out of my Requests would really facilitate this type of testing. Oh, and I'd love to be able to run such a test suite in a CI environment as well, but first I'd just like to be able to get it working manually, hehe.

**Reply by Ariemeth** [reply]
You could use a conditional block after the request to check that the user's name exists in the list, then another conditional to check it isn't there after a different request.  You could add a teardown request that gets called if either of the previous checked conditionals fail as well as at the end

**Reply by ShawnFUB** [reply]
Sure - the flow itself isn't necessarily the problem - it's asserting a test pass/failure as the flow takes the various branches. The Test Summary block exists and works well, but the only way to emit test pass/fail data to it is from Request Tests and there's no way for Flow to do anything equivalent as far as I can tell.

So then I'm back to having multiple copies of my Request for each set of Test assertions because the assertions are specific to where we are in the workflow rather than the Requests themselves.

Example: if I `GET person/id` for a person who doesn't exist, I _want_ a 403 response code - that's not a test failure, it's the proper API behavior. But I don't want to bake that into my request definition because the expected behavior depends on the state of the API, and that depends on the calls that precede it.

Flows could resolve this issue I've had with using Postman for automated testing - but the missing piece is an Assertion block that accepts request/response Data and emits a test result that could be routed to the Test Summary block.

**Reply by Ariemeth** [reply]
I'm suggesting dropping the test script and handle the testing cases in flows directly.
It would be nice if there was more you could do with the test output of a request.

**Comment by cahan1013** [comment]
I'm using Flows for e2e integration testing invoking APIs.  Seriously, it would be appreciated that Flows can provide features below:
1. Copy-Paste. I can not copy part of flows to another  and need to re-create very similar steps here and there ,which is a waste of time.... 
2. Disable some of steps in a flow. it will be very helpful especially if i only want to test like first 3/5 steps while the rest 2/5 steps are still under-development.

**Comment by brineryte** [comment]
I'm the QA Automation engineer for my team and have found flows very useful for running a chain of related authentication requests in order to get a client token that I can then spit out to the console or terminal and use with my automated tests.

**Comment by minoseah629** [comment]
hello, 
Is there way to keep a block in the flow, but have the block disabled.  

postman flow is similar to rapidminer in how blocks are connected. but i am not on my machine where its installed to show it.

<img width="294" alt="image" src="https://user-images.githubusercontent.com/1980068/213775616-61488933-8dee-4ba1-9c1c-4b9693d3afd6.png">

**Reply by minoseah629** [reply]
moving this to new discussion

**Comment by msanchezdev** [comment]
I am using flows to document, well the whole flow of a feature in terms of API calls.

**Features I Long For:**
- Being able to set variables or run a setup script before the flow runs
   Currently it's hard to run the flow multiple times with different data.
   - Along with this, having a mocking library like **faker.js** to be used on the pre-request scripts
   - + Be able to open a dialog or some other prompt for data that needs to be manually inputted for the flow to continue.
- Set up default environment for the whole Flow

![image](https://user-images.githubusercontent.com/35346082/222216577-84a93c29-9043-4f19-bb7d-7e9aa5863a92.png)

**Reply by msanchezdev** [reply]
**Bug I Found:**
- It suddendly deleted all my other nodes when I removed the environment on the first "Send Request" node, and I confirmed I wanted to update the environment on all 6 nodes.

![image](https://user-images.githubusercontent.com/35346082/222217932-ce579bbb-f07f-44d0-b74f-2084a1b58147.png)
